0003-secret-seed-storage-and-key-derivation.md

```diff
@@ -0,0 +1,201 @@
+
+## Secret Seed Storage
+
+This section recommends how we should generate and store the secret seed by answering key questions.
+
+### Can you recover anything if you have your secret seed but no database backup?
+
+**In practice, No.**
+
+In Theory and in the case of Ethereum only you could scan the blockchain for any HTLC looking contract paying money to an address you generated deterministically from your seed and reclaim it.
+
+In Bitcoin you can't because you'll never be able to compute the P2WSH scriptPubKey without knowing the other person's input.
+If it has already been redeemed, you could scan for redeem transactions with script that matches our Bitcoin HTLC, recover the secret and then look for the HTLC on the other chain (as above, assuming it's Ethereum).
+
+I don't think this partial recovery is worth putting the significant effort that it would require to implement.
+
+### Should it be encrypted?
+
+I think this should be optional.
+A COMIT node doesn't hold funds long term.
+It can only control the funds during the execution of a protocol.
+This is different to LND where if someone gets access to your seed they can take all your funds regardless of whether you have channels open or not.
+
+The best attack vector against an unencrypted seed is for the attacker to get short term access to the system COMIT is running on, covertly steal their secret seed and then wait for them to execute a swap and then redeem to their own address.
+Encrypting the seed does give you protection against this.
+
+Having said that, giving an attacker access to your COMIT node will probably give them access to the HTTP API which they can use to redeem funds to their own address.
+This is probably a better attack vector than the above.
+
+Note: We could prevent this attack by giving the node a way of generating the final redeem addresses without getting them from the api.
+
+For the sake of expediency I propose we incrementally increase the security, i.e. (i) leave the secret unencrypted for now, (ii) make encryption an option later on and (iii) make encryption the default eventually.
+
+### Where to store it?
+
+I don't think we should store secret seed in the database file (this is what lnd does).
+This lets users backup the secret seed separately from the database (although they need to backup both).
+Given that the two are linked, we should store a hash of the secret seed along with the swap id (in the same row or whatever).
+This allows the user to change the secret seed while not invalidating the entire database.
+
+By default I'd put it somewhere like:
+
+```
+~/.comit/secret_seed
+```
+
+But its location should be configurable with `secret_seed_file` in the `.toml` config.
+
+The format of the file should be toml with a single entry `secret_seed` (so we can `encrypted_secret_seed` later).
+
+We should make the permissions restrictive so only the user has read permission on it.
```

Comment:
I assume you mean that if the unix permission are `og+r` then `comit_node` should reject the seed and fail to start? similar to ssh?

Reply:
Interesting idea. I wasn't thinking that actually. Just that when it's first created it should be created with restrictive permissions. It is much less clear in the case of comit what the correct permissions are for the file and I would just let the user decide but provide them with sane defaults.